### PR TITLE
switch webhook configuration from sns to argus

### PR DESCRIPTION
**Pull request by kcajmagic**

### CODE REVIEW

File: caduceus.yaml [hunk 1]
```diff
@@ -343,4 +330,4 @@
   # perSourceIncomingPayloadSizeBuckets:
   #   - 100
   #   - 1000
-  #   - 10000
+  #   - 10000
```
Comment (joe94) [inline caduceus.yaml:333]:
```suggestion
  #   - 10000
  
```
unix text file newline at eof

File: webhookHandler.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	} else {
+		rw.Header().Set("Content-Type", "application/json")
+		rw.Write(msg)
+	}
+}
+
+// update is an api call to processes a listenener registration for adding and updating
+func (r *Registry) UpdateRegistry(rw http.ResponseWriter, req *http.Request) {
+	payload, err := ioutil.ReadAll(req.Body)
+	req.Body.Close()
```
Comment (joe94) [inline webhookHandler.go:92]:
```suggestion
```

The underlying already does this.

File: webhookHandler.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+type RegistryConfig struct {
+	Logger      log.Logger
+	Listener    webhookclient.ListenerFunc
+	ArgusConfig webhookclient.ClientConfig
+}
+
+func NewRegistry(config RegistryConfig, listener webhookclient.Listener) (*Registry, error) {
+	argus, err := webhookclient.CreateClient(config.ArgusConfig, webhookclient.WithLogger(config.Logger))
```
Comment (joe94) [inline webhookHandler.go:27]: where are we registering the config.Listener (caduceusSenderWrapper.Update in this case)?

File: caduceus.yaml [hunk 4]
```diff
@@ -230,59 +230,46 @@
...
-  # (Optional)
-  soa:
-    provider: "example.awsdns-19.com:53"
+  authHeader: ["xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx=","dXNlcjpwYXNz"]
 
 ########################################
 #   Webhook Related Configuration
 ########################################
+  # Argus Config for storing the webhook information
+  argusConfig:
```
Comment (joe94) [inline caduceus.yaml:239]: Should we make the toplevel field `webhook` here and place all the webhook related configs under it? Then we can do the same for scytale and keep both consistent.

File: caduceus.yaml [hunk 5]
```diff
@@ -230,59 +230,46 @@
...
-  # (Optional)
-  soa:
-    provider: "example.awsdns-19.com:53"
+  authHeader: ["xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx=","dXNlcjpwYXNz"]
 
 ########################################
 #   Webhook Related Configuration
 ########################################
+  # Argus Config for storing the webhook information
+  webhookConfig:
```
Comment (joe94) [inline caduceus.yaml:239]: Why not just `webhook`?

File: main.go [hunk 6]
```diff
@@ -138,33 +132,40 @@ func caduceus(arguments []string) int {
...
+		measures.WebhookListSize.Set(float64(len(items)))
 	}
-	webhookRegistry, webhookHandler := webhookFactory.NewRegistryAndHandler(metricsRegistry)
-	webhookFactory.SetExternalUpdate(caduceusSenderWrapper.Update)
+	webhookRegistry, err := NewRegistry(RegistryConfig{
+		Logger: logger,
+		Listener: func(items []model.Item) {
+			hooks := make([]webhook.W, len(items))
+			for index, item := range items {
+				hook, err := convertItemToWebhook(item)
```
Comment (joe94) [inline main.go:146]: if conversions fail, hooks will have holes with nil objects in them and the Update() function will panic https://github.com/xmidt-org/caduceus/blob/master/senderWrapper.go#L152
We should probably just append successfully converted objects and log error cases

File: caduceus.yaml [hunk 7]
```diff
@@ -230,59 +230,46 @@
...
-  # can restart much faster.
-  # (Optional)
-  soa:
-    provider: "example.awsdns-19.com:53"
+  authHeader: ["xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx=","dXNlcjpwYXNz"]
 
 ########################################
 #   Webhook Related Configuration
 ########################################
+  # Argus Config for storing the webhook information
```
Comment (kristinaspring) [inline caduceus.yaml:238]:
```suggestion
  # webhookStore provides configuration for storing and obtaining webhook 
  # information using argus.
```

File: caduceus_type.go [hunk 8]
```diff
@@ -34,6 +35,7 @@ type CaduceusConfig struct {
 	JobQueueSize     int
 	Sender           SenderConfig
 	JWTValidators    []JWTValidator
+	WebhookConfig    chrysom.ClientConfig
```
Comment (kristinaspring) [inline caduceus_type.go:38]:
```suggestion
	WebhookStore    chrysom.ClientConfig
```

File: deploy/packaging/caduceus_spruce.yaml [hunk 9]
```diff
@@ -235,64 +235,18 @@ jwtValidators:
...
-    region: (( grab $AWS_REGION || "us-east-1" ))
-
-    # topicArn describes the SNS topic this server needs to subscribe to
-    topicArn: (( grab $AWS_SNS_ARN || "arn:aws:sns:us-east-1:000000000000:xmidt-local-caduceus"))
-
-    #urlPath is the URL path SNS will use to confirm a subscription with this server
-    urlPath: "/api/v2/aws/sns"
-
+# Argus Config for storing the webhook information
+webhookConfig:
```
Comment (kristinaspring) [inline deploy/packaging/caduceus_spruce.yaml:244]:
```suggestion
webhookStore:
```

File: caduceus.yaml (hunk 5 above)
Comment (kristinaspring) [inline caduceus.yaml:239]:
```suggestion
  webhookStore:
```

File: main.go [hunk 10]
```diff
@@ -82,13 +82,7 @@ func caduceus(arguments []string) int {
...
 	}
 
-	var (
-		infoLog  = log.WithPrefix(logger, level.Key(), level.InfoValue())
-		errorLog = log.WithPrefix(logger, level.Key(), level.ErrorValue())
-		debugLog = log.WithPrefix(logger, level.Key(), level.DebugValue())
-	)
-
-	infoLog.Log("configurationFile", v.ConfigFileUsed())
+	logging.Info(logger).Log("configurationFile", v.ConfigFileUsed())
```
Comment (kristinaspring) [inline main.go:85]: I just got rid of all these `logging.Info()` in caduceus to increase performance - please use
```
log.WithPrefix(logger, level.Key(), level.InfoValue()).Log(whatever...)
``` 
or add the level.Key() stuff to the log statement itself.

File: main.go [hunk 11]
```diff
@@ -138,33 +132,40 @@ func caduceus(arguments []string) int {
...
-	webhookRegistry, webhookHandler := webhookFactory.NewRegistryAndHandler(metricsRegistry)
-	webhookFactory.SetExternalUpdate(caduceusSenderWrapper.Update)
+	webhookRegistry, err := NewRegistry(RegistryConfig{
+		Logger: logger,
+		Listener: func(items []model.Item) {
+			hooks := make([]webhook.W, len(items))
+			for index, item := range items {
+				hook, err := convertItemToWebhook(item)
+				if err != nil {
+					continue
```
Comment (kristinaspring) [inline main.go:148]: we should probably log this or put it in a metric or something - we don't expect any failures here and suppressing this error could make it more difficult to debug later.

File: main.go [hunk 12]
```diff
@@ -203,23 +191,7 @@ func caduceus(arguments []string) int {
...
-		errorLog.Log(logging.ErrorKey(), webhookStartResults.Error)
-	} else {
-		// todo: add message
-		webhookFactory.SetList(webhook.NewList(webhookStartResults.Hooks))
-		caduceusSenderWrapper.Update(webhookStartResults.Hooks)
-	}
-
-	debugLog.Log(messageKey, "Current listener retrieval.", "elapsedTime", time.Since(beginObtainList))
-	infoLog.Log(messageKey, "Caduceus is up and running!", "elapsedTime", time.Since(beginCaduceus))
+	logging.Info(logger).Log(logging.MessageKey(), "Caduceus is up and running!", "elapsedTime", time.Since(beginCaduceus))
```
Comment (kristinaspring) [inline main.go:194]: This also shouldn't use `logging.Info()`.

File: webhookHandler.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+// jsonResponse is an internal convenience function to write a json response
+func jsonResponse(rw http.ResponseWriter, code int, msg string) {
+	rw.Header().Set("Content-Type", "application/json")
+	rw.WriteHeader(code)
+	rw.Write([]byte(fmt.Sprintf(`{"message":"%s"}`, msg)))
+}
+
+// get is an api call to return all the registered listeners
+func (r *Registry) GetRegistry(rw http.ResponseWriter, req *http.Request) {
```
Comment (kristinaspring) [inline webhookHandler.go:50]: This is a change to the Caduceus API - previously, we were returning the registry that Caduceus has stored locally.  Now, we are returning the list of webhooks that Argus gives us, but not the list that Caduceus is currently working with.  A Get() on chrysom also doesn't even update the listener, so Caduceus could actually return a list with webhooks it doesn't have yet.

**Pull request merged**